README.md
# logstash_lwrp chef cookbook

[![CircleCI](https://circleci.com/gh/petracvv/logstash_lwrp/tree/master.svg?style=svg)](https://circleci.com/gh/petracvv/logstash_lwrp/tree/master) [![Cookbook Version](https://img.shields.io/cookbook/v/logstash_lwrp.svg)](https://supermarket.chef.io/cookbooks/logstash_lwrp)

A resource-driven [Chef](http://chef.io) cookbook for installing and running [Logstash](http://elastic.co/products/logstash) on GNU/Linux systems.

Provides resources for installing, configuring and running the [currently supported versions](https://www.elastic.co/support/eol) of the Logstash application.

**Installation of Logstash prerequisites such as Java are out-of-scope for this cookbook**

## Supported Software / Platforms

### Platforms

- Debian / Ubuntu
- RHEL / CentOS

### Chef

- Chef 13+

### Logstash

- 5.x
- 6.x
- 7.x

## Usage

This cookbook provides resources to manage Logstash including ones to install the application, create base configuration, setup processing pipelines and managing the service. In general, a normal deployment will need to use one of each resource to have a fully functioning Logstash install.

### Example

```ruby
logstash_install 'example' do
  action :install
end

logstash_config 'example' do
  action :create
end

logstash_pipeline 'main' do
  instance 'example'
  config_string "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
  pipeline_workers 1
  action :create
end

logstash_service 'example' do
  xms '256M'
  xmx '256M'
  action [:start, :enable]
end
```

More complex examples can be found in the test cookbook recipe: https://github.com/petracvv/logstash_lwrp/blob/master/test/fixtures/cookbooks/logstash_lwrp_test/recipes/default.rb

## Resources

### logstash_install

logstash_install installs an instance of the Logstash application using the official tar package found on Elastic's main site. At the moment, only one instance of Logstash per node is officially supported but it is possible to have multiple instances if they are using different Logstash versions.

#### Actions

- `:install`: Installs an instance of the logstash application
- `:upgrade`: Upgrades an instance of the logstash application

#### Properties

|property|description|default|kind_of|
|--------|-----------|-------|-------|
|instance|Name of the installation|nil|String|
|version|Logstash version to install|`'7.1.0'`|String|
|ls_user|User Logstash will run under|`'logstash'`|String|
|ls_group|Group Logstash will run under|`'logstash'`|String|
|ls_shell|Shell assigned to Logstash user|`'/sbin/nologin'`|String|
|install_path|Where to extract the logstash tarball|`'/opt/logstash_INSTANCENAME_VERSION/'`|String|
|install_mode|Permissions for install directory|`'0750'`|String|
|tarball_uri|Location of Logstash tarball|`'https://artifacts.elastic.co/downloads/logstash/logstash-VERSION.tar.gz'`|String|
|checksum_uri|Location of checksum for Logstash tarball|`'https://artifacts.elastic.co/downloads/logstash/logstash-VERSION.tar.gz.sha512'`|String|

#### Examples

Install a Logstash 7.1.0 instance named 'testing' to `/opt/logstash_testing_7_1_0/` with a symlink at `/opt/logstash_testing`

```ruby
logstash_install 'testing' do
  version '7.1.0'
  action :install
end
```

### logstash_config

Installs the base Logstash configuration into the `logstath.yml` file. **If you are using a Logstash 5.x version without multiple pipeline support, you will need to add your pipeline configuration with this resource instead of the `logstash_pipeline` resource.**

#### Actions

- `:create`: Deploys the configuration to the `logstash.yml` file of the named Logstash instance

#### Properties

|property|description|default|kind_of|
|--------|-----------|-------|-------|
|instance|Name of the installation|nil|String|
|mode|Permission mode of the `logstash.yml` file|`'0640'`|String|
|node_name|Name of the logstash node|`node['hostname']`|String|
|path_data|Location of persistent Logstash data|`'/opt/logstash_INSTANCE/data'`|String|
|path_logs|Location of Logstash log files|`'/opt/logstash_INSTANCE/logs'`|String|
|log_level|Log level of Logstash process|`'info'`|String|
|log_format|Log format of Logstash process|`'plain'`|String|
|pipeline_defaults|Default values for pipelines|nil|Hash|

#### Examples

Create a logstash configuration with a custom node name for the 'testing' instance

```ruby
logstash_config 'testing' do
  node_name 'logstash-test'
  action :create
end
```

Deploy the main pipeline for a 'testing' Logstash 5.x install. **This should only be done in Logstash 5.x installs since it does not support pipelines.yml**

```ruby
logstash_config 'testing' do
  node_name 'logstash-5-test'
  pipeline_defaults(
    'pipeline.id' => 'main',
    'path.config' => '/etc/mycustomlogstash/'
  )
  action :create
end
```

### logstash_pipeline

This resource deploys a pipeline to the `pipelines.yml` config file for Logstash versions that support it.

There are two main ways to deploy pipeline configuration in Logstash:

- Using a `config.string` where all your config is defined in one string in pipelines.yml
- Using a `path.config` where logstash looks for configuration files in the target path specified.

If supplied with a `config_string` property, `logstash_pipeline` will deploy a pipeline with the configuration in the `config.string` logstash property

If supplied with a `config_templates` property, `logstash_pipeline` will create a per-pipeline directory in the Logstash install location and deploy the logstash configuration files specified in the property. It will then set the `path.config` Logstash property to the created directory. **The specified templates are always deployed from the wrapper cookbook (i.e. where the `logstash_pipeline` resource is defined)**

#### Actions

- `:create`: Deploy a the pipelines.yml with the pipeline config specified

#### Properties

|property|description|default|kind_of|
|--------|-----------|-------|-------|
|pipeline_id|Name of the pipeline|nil|String|
|instance|Name of the installation|nil|String|
|mode|Permission mode of the pipelines.yml file|`'0640'`|String|
|pipeline_workers|Number of threads for this pipeline|`node['cpu']['cores']`|Integer|
|config_string|Logstash pipeline configuration in one string|nil|String|
|config_templates|Array of template names to deploy as the Logstash pipeline configuration|nil|Array|
|config_template_variables|Hash of template variables to use in the Logstash pipeline templates|nil|Hash|
|pipeline_settings|Other pipeline configuration options|nil|Hash|

#### Examples

Deploy a short pipeline configuration for the 'testing' Logstash install.

```ruby
logstash_pipeline 'string' do
  instance 'testing'
  pipeline_workers 1
  config_string "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
  action :create
end
```

Deploy a more complicated Logstash pipeline for the 'testing' Logstash instance using templates and variables in your wrapper cookbook

```ruby
logstash_pipeline 'main' do
  instance 'testing'
  config_templates %w( input_syslog.conf.erb output_elasticsearch.conf.erb output_stdout.conf.erb)
  config_template_variables(
    'test.id' => '321',
    'test.name' => 'Variables'
  )
  action :create
end
```

### logstash_service

This resource sets up the Logstash system service and sets the runtime Java options for the service. The service is normally generated using a provided upstream script. If you want to use your own service definition, at the moment only custom systemd services are supported by using the `systemd_unit_content` property.

#### Actions

- `:start`: Creates the service and starts it
- `:stop`: Stops the service
- `:restart`: Restarts the service
- `:enable`: Creates the service and enables it (starts on boot)

#### Properties
|property|description|default|kind_of|
|--------|-----------|-------|-------|
|instance|Name of the Logstash installation|nil|String|
|service_name|Name of the Logstash service|`'logstash_INSTANCE'`|String|
|systemd_unit_content|Content of systemd unit file. Will override any autogenerated service definition|nil|String|
|description|Longer description of the service|`'logstash'`|String|
|javacmd|Path to java binary|`'/usr/bin/java'`|String|
|ls_settings_dir|Directory with logstash.yml file|`'/opt/logstash_INSTANCE/config'`|String|
|ls_opts|Command line arguments to Logstash|`['--path.settings LS_SETTINGS_DIR']`|Array|
|ls_pidfile|Path to Logstash pidfile (only relevant for sysv)|`'/var/run/SERVICE_NAME.pid'`|String|
|ls_gc_log_file|Path to Logstash log file for Java GC|`'/opt/logstash_INSTANCE/logs/SERVICE_NAME_gc.log'`|String|
|ls_open_files|Open file limit for Logstash service|16384|Integer|
|ls_nice|Nice value for logstash process|19|Integer|
|ls_prestart|Command or script to run before Logstash service starts|nil|String|
|xms|Java Xms property for Logstash service|`node['memory'] ? "#{(node['memory']['total'].to_i * 0.6).floor / 1024}M" : '1G'`|String|
|xmx|Java Xmx property for Logstash service|`node['memory'] ? "#{(node['memory']['total'].to_i * 0.6).floor / 1024}M" : '1G'`|String|
|gc_opts|Java garbage collection options for Logstash service|see resource definition|Array|
|java_opts|Other Java options for Logstash resource|see resource definition|Array|

#### Examples

Install and start the Logstash service for the 'testing' instance with 256M as the memory limit

```ruby
logstash_service 'testing' do
  xms '256M'
  xmx '256M'
  action [:start, :enable]
end
```

## Acknowledgements

Thanks to:

- The [tomcat cookbook](https://github.com/chef-cookbooks/tomcat) authors for inspiration for the logstash_install resource pattern.
- The [haproxy cookbook](https://github.com/sous-chefs/haproxy) authors for great examples of the Chef accumulator pattern

## License & Authors

- Author: Mihai Petracovici (https://github.com/petracvv)
  
```
Copyright:: 2018, Mihai Petracovici

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# logstash_lwrp Cookbook CHANGELOG

This file is used to list changes made in each version of the logstash_lwrp cookbook.

## Unreleased

## v2.1.0 (2021-01-24)

- Fix for YAML not being loaded by default in newer chef 16 versions
- Remove Ubuntu 16.04 as a tested platform (EOL in April)

## v2.0.0 (2020-07-21)

- Update to support Chef 16
- Bump required chef version to v15.8

## v1.2.0 (2019-06-10)

- Update to support Logstash 7.x
- Add ChefSpec test coverage
- Add systemd_unit_content property to logstash_service for custom systemd unit config
- Update to latest Circle CI test workflow

## v1.1.0 (2019-02-17)

- Add template variables property for logstash_pipeline.
- Add Circle CI tests for all supported platforms.

## v1.0.1 (2018-11-18)

- Add CONTRIBUTING.md file
- Reformat README so rendering works on github and supermarket

## v1.0.0 (2018-11-18)

Initial release.

- Added logstash_install resource
- Added logstash_config resource
- Added logstash_pipeline resource
- Added logstash_service resource
- Added Initial README
- Added LICENSE
- Added Integration tests


spec/config_spec.rb
require 'spec_helper'

describe 'logstash_config' do
  step_into :logstash_config
  platform 'ubuntu'
  automatic_attributes['hostname'] = 'logstash-test'

  context 'with default options' do
    recipe do
      logstash_install 'test'
      logstash_config 'test'
      logstash_service 'test'
    end

    it do
      is_expected.to create_template('/opt/logstash_test/config/logstash.yml').with(
        user: 'logstash',
        group: 'logstash',
        mode: '0640'
      )
    end

    it do
      is_expected.to render_file('/opt/logstash_test/config/logstash.yml').with_content { |content|
        expect(content).to include('node.name: logstash-test')
        expect(content).to include('path.data: /opt/logstash_test/data')
        expect(content).to include('path.logs: /opt/logstash_test/logs')
        expect(content).to include('log.level: info')
        expect(content).to include('log.format: plain')
      }
    end
  end
end


spec/install_spec.rb
require 'spec_helper'

describe 'logstash_install' do
  step_into :logstash_install
  platform 'ubuntu'

  automatic_attributes['hostname'] = 'logstash-test'

  before do
    allow(Dir).to receive(:empty?).and_call_original
    allow(Dir).to receive(:empty?).with('/opt/logstash_test_7_1_0/').and_return(true)
  end

  context 'with default options' do
    recipe do
      logstash_install 'test'
      logstash_config 'test'
      logstash_service 'test'
    end

    it { is_expected.to install_package('tar') }

    it { is_expected.to create_group('logstash') }

    it do
      is_expected.to create_user('logstash').with(
        gid: 'logstash',
        shell: '/sbin/nologin',
        system: true
      )
    end

    it do
      is_expected.to create_directory('/opt/logstash_test_7_1_0/').with(
        owner: 'logstash',
        group: 'logstash',
        mode: '0750',
        recursive: true
      )
    end

    it { is_expected.to create_remote_file("#{Chef::Config['file_cache_path']}/logstash-7.1.0.tar.gz") }

    it { expect(chef_run.remote_file("#{Chef::Config['file_cache_path']}/logstash-7.1.0.tar.gz")).to notify('execute[extract logstash tarball]').to(:run).immediately }

    it { expect(chef_run.execute('chown logstash installation as logstash')).to subscribe_to('execute[extract logstash tarball]').on(:run).immediately }

    it { is_expected.to create_link('/opt/logstash_test').with_to('/opt/logstash_test_7_1_0/') }
  end

  context 'with upgrade action' do
    before do
      allow(File).to receive(:symlink?).and_call_original
      allow(File).to receive(:symlink?).with('/opt/logstash_test').and_return(true)

      allow(File).to receive(:realpath).and_call_original
      allow(File).to receive(:realpath).with('/opt/logstash_test').and_return('/opt/logstash_test_6_8_0')
    end

    recipe do
      logstash_install 'test' do
        action :upgrade
      end
      logstash_config 'test'
      logstash_service 'test'
    end

    it { expect(chef_run.notify_group('stop logstash service before upgrade')).to notify('logstash_service[test]').to(:stop).immediately }

    it { is_expected.to run_ruby_block('Install new version') }

    it do
      is_expected.to run_execute('copy persistent data from old install').with(
        command: 'cp -a /opt/logstash_test_6_8_0/data /opt/logstash_test/',
        user: 'logstash'
      )
    end
  end
end


spec/service_spec.rb
require 'spec_helper'

describe 'logstash_service' do
  step_into :logstash_service
  automatic_attributes['memory']['total'] = '1024'

  recipe do
    logstash_install 'test'
    logstash_service 'test'
  end

  # Focus on systemd for tests
  platform 'ubuntu'

  before do
    allow(Chef::Platform::ServiceHelpers).to receive(:service_resource_providers).and_call_original
    allow(Chef::Platform::ServiceHelpers).to receive(:service_resource_providers).and_return([:systemd])
  end

  context 'when starting service' do
    it do
      is_expected.to create_template('/opt/logstash_test/config/jvm.options').with(
        owner: 'logstash',
        group: 'logstash',
        mode: '0640'
      )
    end

    it do
      is_expected.to render_file('/opt/logstash_test/config/jvm.options').with_content { |content|
        expect(content).to include('Xms614')
        expect(content).to include('Xmx614')
        expect(content).to include('-XX:+UseConcMarkSweepGC')
        expect(content).to include('-XX:CMSInitiatingOccupancyFraction=75')
        expect(content).to include('-XX:+UseCMSInitiatingOccupancyOnly')
        expect(content).to include('-Djava.awt.headless=true')
        expect(content).to include('-Dfile.encoding=UTF-8')
        expect(content).to include('-Djruby.compile.invokedynamic=true')
        expect(content).to include('-Djruby.jit.threshold=0')
        expect(content).to include('-XX:+HeapDumpOnOutOfMemoryError')
        expect(content).to include('-Djava.security.egd=file:/dev/urandom')
      }
    end

    context 'with upstream unit file' do
      it do
        is_expected.to create_template('/opt/logstash_test/config/startup.options').with(
          owner: 'logstash',
          group: 'logstash',
          mode: '0640'
        )
      end

      it do
        is_expected.to render_file('/opt/logstash_test/config/startup.options').with_content { |content|
          expect(content).to include('JAVACMD=/usr/bin/java')
          expect(content).to include('LS_HOME=/opt/logstash_test')
          expect(content).to include('LS_SETTINGS_DIR=/opt/logstash_test/config')
          expect(content).to include('LS_OPTS="--path.settings /opt/logstash_test/config"')
          expect(content).to include('LS_PIDFILE=/var/run/logstash_test.pid')
          expect(content).to include('LS_USER=logstash')
          expect(content).to include('LS_GROUP=logstash')
          expect(content).to include('LS_GC_LOG_FILE=/opt/logstash_test/logs/logstash_test_gc.log')
          expect(content).to include('LS_OPEN_FILES=16384')
          expect(content).to include('LS_NICE=19')
          expect(content).to include('SERVICE_NAME="logstash_test"')
          expect(content).to include('SERVICE_DESCRIPTION="logstash"')
        }
      end

      it { expect(chef_run.template('/opt/logstash_test/config/startup.options')).to notify('execute[Generate service file]').to(:run).immediately }
      it { expect(chef_run.template('/opt/logstash_test/config/startup.options')).to notify('logstash_service[test]').to(:restart) }
    end

    context 'with custom unit file' do
      recipe do
        logstash_install 'test'
        logstash_service 'test' do
          systemd_unit_content <<-EOU.gsub(/^\s+/, '')
          [Unit]
          Description=logstash

          [Service]
          Type=simpleUser=logstash
          Group=logstash
          # Load env vars from /etc/default/ and /etc/sysconfig/ if they exist.
          # Prefixing the path with '-' makes it try to load, but if the file doesn't
          # exist, it continues onward.
          EnvironmentFile=-/etc/default/logstash_test
          EnvironmentFile=-/etc/sysconfig/logstash_test
          ExecStart=/opt/logstash_test/bin/logstash "--path.settings" "/opt/logstash_test/config"
          Restart=always
          WorkingDirectory=/
          Nice=19
          LimitNOFILE=16384

          [Install]
          WantedBy=multi-user.target
          EOU
        end
      end

      it { is_expected.to create_systemd_unit('logstash_test.service') }
    end

    it { is_expected.to start_service('logstash_test').with(provider: Chef::Provider::Service::Systemd) }
  end

  context 'when stopping service' do
    recipe do
      logstash_install 'test'
      logstash_service 'test' do
        action :stop
      end
    end
    it { is_expected.to stop_service('logstash_test') }
  end

  context 'when restarting service' do
    recipe do
      logstash_install 'test'
      logstash_service 'test' do
        action :restart
      end
    end
    it { is_expected.to restart_service('logstash_test') }
  end

  context 'when enabling service' do
    recipe do
      logstash_install 'test'
      logstash_service 'test' do
        action :enable
      end
    end
    it { is_expected.to enable_service('logstash_test') }
  end
end


resources/config.rb
#
# Cookbook:: logstash_lwrp
# Resource:: config
#
# Copyright:: 2018, Mihai Petracovici
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
resource_name :logstash_config
provides :logstash_config

property :instance, String, name_property: true
property :mode, String, default: '0640'
property :node_name, String, default: lazy { node['hostname'] }
property :path_data, String, default: lazy { |r| "/opt/logstash_#{r.instance}/data" }
property :path_logs, String, default: lazy { |r| "/opt/logstash_#{r.instance}/logs" }
property :log_level, String, default: 'info', equal_to: %w( info fatal error warn debug trace )
property :log_format, String, default: 'plain', equal_to: %w( plain json )
property :pipeline_defaults, Hash

action :create do
  template "#{logstash_home}/config/logstash.yml" do
    source 'config/logstash.yml.erb'
    owner logstash_user
    group logstash_group
    mode new_resource.mode
    cookbook 'logstash_lwrp'
    variables(
      node_name: new_resource.node_name,
      path_data: new_resource.path_data,
      log_level: new_resource.log_level,
      log_format: new_resource.log_format,
      path_logs: new_resource.path_logs,
      pipeline_defaults: new_resource.pipeline_defaults
    )
    action :create
    notifies :restart, "logstash_service[#{new_resource.instance}]", :delayed
  end
end

action_class do
  include LogstashLWRP::Helpers
end


libraries/helpers.rb
#
# Cookbook:: logstash_lwrp
# Module:: Helpers
#
# Copyright:: 2018, Mihai Petracovici
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module LogstashLWRP
  module Helpers
    def logstash_home
      "/opt/logstash_#{new_resource.instance}"
    end

    def logstash_user
      find_resource(:logstash_install, new_resource.instance).ls_user
    end

    def logstash_group
      find_resource(:logstash_install, new_resource.instance).ls_group
    end
  end
end


resources/install.rb
#
# Cookbook:: logstash_lwrp
# Resource:: install
#
# Copyright:: 2018, Mihai Petracovici
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Resource based on install resource from official tomcat cookbook.
# https://github.com/chef-cookbooks/tomcat/blob/master/resources/install.rb

resource_name :logstash_install
provides :logstash_install

property :instance, String, name_property: true
property :version,  String, default: '7.1.0', callbacks: {
  'should be in X.Y.Z format' => lambda { |v|
    v =~ /\d+.\d+.\d+/
  },
}
property :ls_user,  String, default: 'logstash'
property :ls_group, String, default: 'logstash'
property :ls_shell, String, default: '/sbin/nologin', callbacks: {
  'should be an installed shell' => lambda { |s|
    ::File.foreach('/etc/shells').grep s
  },
}
property :install_path, String, default: lazy { |r| "/opt/logstash_#{r.instance}_#{r.version.tr('.', '_')}/" }
property :install_mode, String, default: '0750'
property :tarball_uri, String, default: lazy { |r| "https://artifacts.elastic.co/downloads/logstash/logstash-#{r.version}.tar.gz" }
property :checksum_uri, String, default: lazy { |r| "https://artifacts.elastic.co/downloads/logstash/logstash-#{r.version}.tar.gz.sha512" }

action :install do
  package 'tar'

  tarball_path = "#{Chef::Config['file_cache_path']}/logstash-#{new_resource.version}.tar.gz"

  group new_resource.ls_group do
    append true
    action :create
  end

  user new_resource.ls_user do
    gid new_resource.ls_group
    shell new_resource.ls_shell
    system true
    action :create
  end

  directory 'logstash install directory' do
    path new_resource.install_path
    owner new_resource.ls_user
    group new_resource.ls_group
    mode new_resource.install_mode
    recursive true
    action :create
  end

  remote_file "logstash #{new_resource.version} tarball" do
    source new_resource.tarball_uri
    path tarball_path
    verify { |tarball| validate_checksum(tarball) }
    only_if { ::Dir.empty?(new_resource.install_path) } # Avoid redownloading entire logstash tarball every chef run
    action :create
    notifies :run, 'execute[extract logstash tarball]', :immediately
  end

  execute 'extract logstash tarball' do
    command "tar -xzf #{tarball_path} -C #{new_resource.install_path} --strip-components=1"
    creates ::File.join(new_resource.install_path, 'LICENSE.txt')
    action :nothing
  end

  execute "chown logstash installation as #{new_resource.ls_user}" do
    command "chown -R #{new_resource.ls_user}:#{new_resource.ls_group} #{new_resource.install_path}"
    action :nothing
    subscribes :run, 'execute[extract logstash tarball]', :immediately
  end

  link logstash_home do
    to new_resource.install_path
  end
end

action :upgrade do
  # If there is an existing installation
  if ::File.symlink?(logstash_home)
    oldinstall = ::File.realpath(logstash_home)
    oldversion = oldinstall[/[0-9]+_[0-9]+_[0-9]+/].tr('_', '.')

    notify_group 'stop logstash service before upgrade' do
      action :run
      notifies :stop, "logstash_service[#{new_resource.instance}]", :immediately
      not_if { oldversion.eql?(new_resource.version) }
    end

    # Needs to be in a block or else runs at compile-time
    ruby_block 'Install new version' do
      block do
        action_install
      end
      action :run
    end

    execute 'copy persistent data from old install' do
      command "cp -a #{oldinstall}/data #{logstash_home}/"
      user new_resource.ls_user
      action :run
      not_if { oldversion.eql?(new_resource.version) }
    end
  else # Otherwise just run the install action
    action_install
  end
end

action_class do
  include ::LogstashLWRP::Helpers

  # we have to do this since remote_file expects a checksum as a string
  def fetch_checksum
    uri = URI(new_resource.checksum_uri)
    request = Net::HTTP.new(uri.host, uri.port)
    request.use_ssl = true if uri.to_s.start_with?('https')
    response = request.get(uri)
    if response.code != '200'
      Chef::Log.fatal("Fetching the Logstash tarball checksum at #{uri} resulted in an error #{response.code}")
      raise
    end
    response.body.split(' ')[0]
  rescue => e
    Chef::Log.fatal("Could not fetch the checksum due to an error: #{e}")
    raise
  end

  def validate_checksum(file_to_check)
    desired = fetch_checksum
    # Logstash uses sha512 for their checksums
    actual = Digest::SHA512.hexdigest(::File.read(file_to_check))

    if desired == actual
      true
    else
      Chef::Log.fatal("The checksum of the Logstash tarball on disk (#{actual}) does not match the checksum provided from the mirror (#{desired}). Renaming to #{::File.basename(file_to_check)}.bad")
      ::File.rename(file_to_check, "#{file_to_check}.bad")
      raise
    end
  end
end


resources/pipelines.rb
#
# Cookbook:: logstash_lwrp
# Resource:: pipelines
#
# Copyright:: 2018, Mihai Petracovici
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
resource_name :logstash_pipelines

property :instance, String, name_property: true
property :mode, String, default: '0640'
property :pipelines, Array, default: [], required: true

action :deploy do
  lsuser = logstash_user
  lsgroup = logstash_group

  template "#{logstash_home}/config/pipelines.yml" do
    source 'config/pipelines.yml.erb'
    owner lsuser
    group lsgroup
    mode new_resource.mode
    variables(
      pipelines: new_resource.pipelines
    )
    cookbook 'logstash_lwrp'
    action :create
    notifies :run, 'execute[verify logstash config]', :immediately
    notifies :restart, "logstash_service[#{new_resource.instance}]", :immediately
  end

  execute 'verify logstash config' do
    command "#{logstash_home}/bin/logstash --path.settings #{logstash_home}/config -t"
    user lsuser
    group lsgroup
    action :nothing
  end
end

action_class do
  include LogstashLWRP::Helpers
end


resources/service.rb
#
# Cookbook:: logstash_lwrp
# Resource:: service
#
# Copyright:: 2018, Mihai Petracovici
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
resource_name :logstash_service
provides :logstash_service

property :instance, String, name_property: true
property :service_name, String, default: lazy { |r| "logstash_#{r.instance}" }
property :systemd_unit_content, String
property :description, String, default: 'logstash'
property :javacmd, String, default: '/usr/bin/java'
property :ls_settings_dir, String, default: lazy { |r| "/opt/logstash_#{r.instance}/config" }
property :ls_opts, Array, default: lazy { |r| ["--path.settings #{r.ls_settings_dir}"] }
property :ls_pidfile, String, default: lazy { |r| "/var/run/#{r.service_name}.pid" }
property :ls_gc_log_file, String, default: lazy { |r| "/opt/logstash_#{r.instance}/logs/#{r.service_name}_gc.log" }
property :ls_open_files, Integer, default: 16384
property :ls_nice, Integer, default: 19, callbacks: {
  'should be a valid nice value' => lambda { |n|
    n >= -20 && n <= 19
  },
}
property :ls_prestart, String
property :xms, String, default: node['memory'] ? "#{(node['memory']['total'].to_i * 0.6).floor / 1024}M" : '1G'
property :xmx, String, default: node['memory'] ? "#{(node['memory']['total'].to_i * 0.6).floor / 1024}M" : '1G'
property :gc_opts, Array, default: ['-XX:+UseConcMarkSweepGC',
                                    '-XX:CMSInitiatingOccupancyFraction=75',
                                    '-XX:+UseCMSInitiatingOccupancyOnly']
property :java_opts, Array, default: ['-Djava.awt.headless=true',
                                      '-Dfile.encoding=UTF-8',
                                      '-Djruby.compile.invokedynamic=true',
                                      '-Djruby.jit.threshold=0',
                                      '-XX:+HeapDumpOnOutOfMemoryError',
                                      '-Djava.security.egd=file:/dev/urandom']

action :start do
  create_init

  service new_resource.service_name do
    supports restart: true, status: true
    provider service_provider
    action :start
  end
end

action :stop do
  service new_resource.service_name do
    supports status: true
    provider service_provider
    action :stop
  end
end

action :restart do
  service new_resource.service_name do
    supports status: true
    provider service_provider
    action :restart
  end
end

action :enable do
  create_init

  service new_resource.service_name do
    supports status: true
    provider service_provider
    action :enable
  end
end

action_class do
  include LogstashLWRP::Helpers

  def service_provider
    available = Chef::Platform::ServiceHelpers.service_resource_providers

    # Service providers in same order of preference as the logstash generator
    if available.include?(:systemd) || property_is_set?(:systemd_unit_content)
      Chef::Provider::Service::Systemd
    elsif available.include?(:upstart)
      Chef::Provider::Service::Upstart
    elsif available.include?(:debian)
      Chef::Provider::Service::Init::Debian
    elsif available.include?(:redhat)
      Chef::Provider::Service::Init::Redhat
    else
      Chef::Log.fatal('Unsupported init system')
      raise
    end
  end

  def create_init
    template "#{new_resource.ls_settings_dir}/jvm.options" do
      source 'init/jvm.options.erb'
      owner logstash_user
      group logstash_group
      cookbook 'logstash_lwrp'
      variables(
        xms: new_resource.xms,
        xmx: new_resource.xmx,
        gc_opts: new_resource.gc_opts,
        java_opts: new_resource.java_opts
      )
      mode '0640'
      action :create
    end

    if property_is_set?(:systemd_unit_content)

      systemd_unit "#{new_resource.service_name}.service" do
        content new_resource.systemd_unit_content
        action :create
      end

    else
      template "#{new_resource.ls_settings_dir}/startup.options" do
        source 'init/startup.options.erb'
        owner logstash_user
        group logstash_group
        mode '0640'
        cookbook 'logstash_lwrp'
        variables(
          javacmd: new_resource.javacmd,
          ls_home: logstash_home,
          ls_settings_dir: new_resource.ls_settings_dir,
          ls_opts: new_resource.ls_opts,
          ls_pidfile: new_resource.ls_pidfile,
          ls_user: logstash_user,
          ls_group: logstash_group,
          ls_gc_log_file: new_resource.ls_gc_log_file,
          ls_open_files: new_resource.ls_open_files,
          ls_nice: new_resource.ls_nice,
          ls_service_name: new_resource.service_name,
          ls_service_description: new_resource.description,
          ls_prestart: new_resource.ls_prestart
        )
        action :create
        notifies :run, 'execute[Generate service file]', :immediately
        notifies :restart, "logstash_service[#{new_resource.instance}]", :delayed
      end

      execute 'Generate service file' do
        command "#{logstash_home}/bin/system-install #{new_resource.ls_settings_dir}/startup.options"
        user 'root'
        action :nothing
      end
    end
  end
end
